gpio: Add a driver for SX1509B

tkln opened the pull request:
Adds a driver for SX1509B I2C GPIO chip. This driver only supports the
basic GPIO features and does not currently implement the LED driver and
keypad matrix features.

Signed-off-by: Aapo Vienamo <aapo.vienamo@iki.fi>

tkln committed:
gpio: Add a driver for SX1509B

Adds a driver for SX1509B I2C GPIO chip. This driver only supports the
basic GPIO features and does not currently implement the LED driver and
keypad matrix features.

Signed-off-by: Aapo Vienamo <aapo.vienamo@iki.fi>

tkln commented:
@MaureenHelm @tbursztyka @anangl Could you please review this?

tbursztyka commented on drivers/gpio/Kconfig.sx1509b:
```diff
@@ -0,0 +1,47 @@
+
+config GPIO_SX1509B_DEV_NAME
+	string "SX1509B GPIO chip Device Name"
+	depends on GPIO_SX1509B
```
you can remove this depends on, since you are already using a condition on that option line 16

tbursztyka commented on drivers/gpio/gpio_sx1509b.c:
```diff
@@ -0,0 +1,338 @@
+	ret = i2c_reg_write_word_be(drv_data->i2c_master, cfg->i2c_slave_addr,
+				    SX1509B_REG_DIR, pins->dir);
+	if (ret)
+		return ret;
```
Use {} even for one liner statements in if/for/while. Apply that everywhere

tbursztyka commented on drivers/gpio/gpio_sx1509b.c:
```diff
@@ -0,0 +1,338 @@
+	switch (access_op) {
+	case GPIO_ACCESS_BY_PIN:
+		if (value) {
+			*pin_data |= 1<<pin;
```
spaces between operands/operators. Apply that everywhere

MaureenHelm commented on drivers/gpio/Kconfig.sx1509b:
```diff
@@ -0,0 +1,44 @@
+	help
+	  Specify the I2C slave address for the SX1509B I2C GPIO chip.
+
+config GPIO_SX1509B_I2C_MASTER_DEV_NAME
```
The device name and address configs should come from dts if possible. Please add bindings in dts/bindings/sensor

MaureenHelm commented on drivers/gpio/gpio_sx1509b.c:
```diff
@@ -0,0 +1,342 @@
+	case GPIO_ACCESS_BY_PIN:
+		if ((flags & GPIO_DIR_MASK) == GPIO_DIR_IN) {
+			pins->dir |= 1 << pin;
+			pins->input_disable &= ~(1 << pin);
```
Use the BIT() macro in misc/util.h

MaureenHelm commented on drivers/gpio/gpio_sx1509b.c:
```diff
@@ -0,0 +1,342 @@
+	default:
+		return -ENOTSUP;
+	}
+
```
need to parse the GPIO_INT flag and return an error if it's set

MaureenHelm commented on drivers/gpio/gpio_sx1509b.c:
```diff
@@ -0,0 +1,342 @@
+	};
+
+	ret = i2c_reg_write_byte(drv_data->i2c_master, cfg->i2c_slave_addr,
+				 SX1509B_REG_RESET, 0x12);
```
Please replace the magic numbers with a more descriptive macro

MaureenHelm reviewed (changes requested):
Do we have any in-tree boards that support this device? Can you please add a build test?

tkln commented:
@MaureenHelm I'm not aware of any current in-tree boards using this, however, I'm working on adding board support for nRF pca20020 (Thingy52) which uses this chip. I'm going to submit the board PR once I'm done with the documentation.

tkln commented:
@MaureenHelm Could you clarify that what you mean by adding a build test? Are there some existing tests or documentation I could refer to?

tkln committed:
tests: drivers: build_all: Add SX1509 GPIO test

A new build test for gpio drivers is added to build_all testcase.yaml.
The gpio_sx1509 driver is enabled in the .conf file and dts.fixup macros
are added for the driver.

Signed-off-by: Aapo Vienamo <aapo.vienamo@iki.fi>

tkln commented:
@MaureenHelm I've now added a build test for the driver.

MaureenHelm commented on drivers/gpio/gpio_sx1509b.c:
```diff
@@ -0,0 +1,372 @@
+	u16_t i2c_slave_addr;
+};
+
+/* General configuratino register addresses */
```
s/configuratino/configuration/

MaureenHelm commented on drivers/gpio/Kconfig.sx1509b:
```diff
@@ -0,0 +1,44 @@
+# Kconfig.pcal9535a - PCAL9535A GPIO configuration options
```
s/pcal9535a/sx1509b/

MaureenHelm reviewed (commented):
Just a couple of minor typos to fix, but otherwise looks good.

tkln commented:
@MaureenHelm Fixed.

MaureenHelm commented on drivers/gpio/gpio_sx1509b.c:
```diff
@@ -0,0 +1,372 @@
+
+	ret = i2c_reg_write_word_be(drv_data->i2c_master, cfg->i2c_slave_addr,
+				    SX1509B_REG_PULL_DOWN,
+				    pins->pull_up);
```
Should the last arg be pins->pull_down?

tkln replied:
Indeed it should.

MaureenHelm commented on drivers/gpio/gpio_sx1509b.c:
```diff
@@ -0,0 +1,372 @@
+
+	ret = i2c_reg_write_word_be(drv_data->i2c_master, cfg->i2c_slave_addr,
+				    SX1509B_REG_OPEN_DRAIN,
+				    pins->pull_up);
```
Should the last arg be pins->open_drain?

tkln replied:
Yes, thanks for pointing that out. Looks like I messed up a merge when working on this earlier.

tkln commented:
@tbursztyka @MaureenHelm The issues you have pointed out should be addressed now.
